Don't use guice for QueryParsers

nik9000 opened the pull request:
1. Gets guice out of the business of building ScoreFunctionParsers and
QueryParsers.
2. Moves QueryParser registration to SearchModule
3. Moves NamedWriteableRegistry construction out of guice and into Node and
TransportClient.
4. Moves shape registration into SearchModule so now all named writeable
registration is done in the SearchModule.

This is breaking for plugin authors. Instead of declaring new QueryParser
like:

```java
public void onModule(IndicesModule module) {
  module.registerQueryParser(NewQueryParser.class);
}
```

you do it like:

```java
public void onModule(SearchModule module) {
  module.registerQueryParser(NewQueryParser::new);
}
```

The QueryParser's argument no longer come from @Inject, now they come from
the declaration in the plugin. The above example is for a no-arg QueryParser.
Most of the QueryParsers in Elasticsearch are no-arg.

ScoreFunctionParsers have a similar but slightly different change. This:

```java
public void onModule(SearchModule module) {
  module.registerFunctionScoreParser(NewFunctionScoreParser.class);
}
```

becomes

```java
public void onModule(SearchModule module) {
  module.registerFunctionScoreParser(new NewFunctionScoreParser());
}
```

Since all known ScoreFunctionParsers have no arg constructors its simpler to
just build them at registration time rather than specify a supplier that is
used to build them later.

javanna commented on core/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -161,14 +237,26 @@
     private final Suggesters suggesters = new Suggesters();
-    private final Set<Class<? extends ScoreFunctionParser>> functionScoreParsers = new HashSet<>();
+    private final Map<String, ScoreFunctionParser<?>> functionScoreParsers = new HashMap<>();
+    private final List<Supplier<QueryParser<?>>> queryParsers = new ArrayList<>();
```
for my education, what is the advantage of using a method reference for query parsers over passing in an actual instance of the query parser? If there is an advantage why don't we do the same for ScoreFunctionParsers too?

nik9000 replied:
b66655113e2015f1b7417707974b1ce15fed8827

javanna replied:
cool thanks for clarifying!

javanna commented on core/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -332,6 +441,87 @@ protected void configureSearch() {
         }
     }
 
+    protected void configureShapes() {
```
can't this be private?

nik9000 replied:
77702e472d1a4651fcb28b3495172acda73e55ff

javanna commented:
looks good! what exactly needs to be discussed here? :) Can we we highlight the changes needed for plugin developers that implement their own query parsers (e.g. just add the code snippet to the description of the issue)?

nik9000 commented:
> looks good! what exactly needs to be discussed here? :)

I dunno. I added it because I wasn't really sure how well it'd be taken. We can't construct all the shared dependencies at the Node level but this one felt ok. We'll need a way for plugins to get access to these dependencies. I was _thinking_ the way to do that is to make them available on the modules in public methods. But that discussion can wait I think.



> Can we we highlight the changes needed for plugin developers that implement their own query parsers (e.g. just add the code snippet to the description of the issue)?

Yeah. I think I can squash and do that.

nik9000 commented:
Squashed and edited the commit description to describe the breaking changes to plugin authors. I've also added `breaking` and `:Plugins` to make it obvious that this is breaking to plugin authors. I've also edited the PR description to include the breaking changes.

javanna commented on core/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -177,8 +274,21 @@ public void registerSuggester(String key, Class<? extends Suggester> suggester)
+        for (String name: parser.getNames()) {
+            Object oldValue = functionScoreParsers.putIfAbsent(name, parser);
+            if (oldValue != null) {
+                throw new IllegalArgumentException("Function score parser [" + oldValue + "] already registered for name [" + name + "]");
```
did you mean name here instead of oldValue?

nik9000 replied:
They are both in the error message. I liked outputting the oldValue because it gives you some hint about what other component registered something with the same name.

javanna commented on core/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -231,16 +342,23 @@ protected void configureFetchSubPhase() {
+            for (String name: parser.names()) {
+                Object oldValue = queryParsersMap.putIfAbsent(name, parser);
+                if (oldValue != null) {
+                    throw new IllegalArgumentException("Query parser [" + oldValue + "] already registered for name [" + name + "] while trying to register [" + parser + "]");
```
did you mean name here instead of oldValue?

javanna replied:
Maybe it's ok, what is oldValue going to print out though?

nik9000 replied:
Standard object toString - just the class with some instanceId. The class part is useful. On the off chance that these things implement a toString I figure we may as well use it.

nik9000 replied:
And by that I mean, the class with instanceId is useful enough that I'll take it. If we get a toString all the better.

nik9000 commented:
@javanna its been a while - does this still look good to you? I'm going to rebase it so we have a chance of being able to merge it.

javanna commented:
sorry @nik9000 this had gotten off my radar. Had another look, LGTM ! cool stuff!

nik9000 commented:
> sorry @nik9000 this had gotten off my radar. Had another look, LGTM ! cool stuff!

Thanks!
